text/0060-enable-creating-and-modifying-of-filescontainers-within-the-safe-browser-environment/0060-enable-creating-and-modifying-of-filescontainers-within-the-safe-browser-environment.md

```diff
@@ -0,0 +1,76 @@
+- Supersedes: (fill me in with a link to RFC this supersedes - if applicable)
+- Superseded by: (fill me in with a link to RFC this is superseded by - if applicable)
+
+## Summary
+
+The NodeJS libraries work via the existing SAFE network CRUD conventions: we take files, we upload them to a network and generate an XORURL we can reference as a result. In the browser we do not have access to a native file sytem, so the underlying NodeJS libraries which the browser calls must allow the uploading of raw bytes, and the creation of empty FilesContainers.
+
+## Conventions
+- The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](http://tools.ietf.org/html/rfc2119).
+
+## Motivation
+
+At the moment, there is no browser native way to create an empty FilesContainer (since the previous convention was the creation of an NFS container which is no longer possible).
+
+## Detailed design
+
+At the moment, the following API calls are available (and implemented in the [NodeJS rust library](https://github.com/maidsafe/safe-nodejs/blob/master/native/src/lib.rs)) with regards to FilesContainers:
+
+* `files_container_create`
+* `files_container_sync`
+* `files_container_get`
+* `files_container_add`
+* `files_container_add_from_raw`
+
+Logic would dictate that the browser would call `files_container_create` to create the files container, however the first argument to this method `location: &str` requires a path to a directory or file which will be uploaded to the FilesContainer. The same issues also exist for `files_container_sync` and `files_container_add`.
+
+`files_container_add_from_raw` is usable, but it always defaults the `type` parameter (which is used as the HTTP `Content-Type` value in the browser to `Raw` regardless of the filetype uploaded..
+
+I propose the modifying of 2 existing APIs in the NodeJS library, and the underlying rust SAFE-API:
+
+* `files_container_create`
+* `files_container_add_from_raw`
+
+#### files_container_create
+
+This method will have its first argument (`location: &str`) made optional, with the default being an empty files container.
+
+#### files_container_add_from_raw
+
+This method should allow the end user to set the `type` of the uploaded file.
+
+In NodeJS pseudocode, the interface would be:
+
+```js
+function files_container_add_from_raw(
+    ArrayBuffer buffer,
+    string url,
+    bool force,
+    bool update_nrs,
+    ?string type,
+    ?bool dry_run
+) : Array;
+```
+
+Setting a type of `text/html` should result in a downloaded `Content-Type` of `text/html`.
+
+The `Content-Type` can also contain a charset value (and any number of mime related data, separated by semi-colons.). Setting a type of `text/html; charset=utf-8` should result in a downloaded `Content-Type` of `text/html; charset=utf-8`
```

Comment:
We don't have support for charset or other type of mime type info in the XOR-URL, therefore I'm not sure we can support these things but just a mime type. Unless we include in the scope of this proposal how to expand XOR-URL encoding.

Reply:
Okay, that's fine, I can amend that easily. It would have been nice to have the hinting but it's a very far away secondary concern compared to the mime type.